File: LC_JS-SOLUTIONS/36. Valid Sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  const rows = Array(9)
    .fill(null)
    .map(() => new Set());
  const cols = Array(9)
    .fill(null)
    .map(() => new Set());
  const square = Array(9)
    .fill(null)
    .map(() => new Set());

  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      let num = board[i][j];

      if (num === ".") {
        continue;
      }

      if (
        rows[i].has(num) ||
        cols[j].has(num) ||
        square[Math.floor(i / 3) * 3 + Math.floor(j / 3)].has(num)
      ) {
        return false;
      }

      rows[i].add(num);
      cols[j].add(num);
      square[Math.floor(i / 3) * 3 + Math.floor(j / 3)].add(num);
    }
  }
  return true;
};

File: LC-Python-Solution/36. Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:

        rows = collections.defaultdict(set)
        cols = collections.defaultdict(set)
        squares = collections.defaultdict(set)

        for r in range(9):
            for c in range(9):
                if board[r][c] == '.':
                    continue
                num = board[r][c]
                if(num in rows[r] or num in cols[c] or num in squares[r//3 , c//3]):
                    return False
                
                rows[r].add(num)
                cols[c].add(num)
                squares[r//3,c//3].add(num)
        
        return True

        


